### PR TITLE
Completed churn-handling for IDM.

### PR DESCRIPTION
**Pull request by Fraser999**
This change is [Review on Reviewable](https://reviewable.io/reviews/maidsafe/safe_vault/367)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @maqi

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/personas/immutable_data_manager.rs [hunk 1]
```diff
@@ -426,94 +431,174 @@ impl ImmutableDataManager {
...
+               data_name,
+               pmid_nodes);
+        // This function is used to filter accounts for which this node is no longer responsible, so
+        // return `None` in this case
+        let close_group = if let Some(group) = self.close_group_to(routing_node, &data_name) {
+            group
+        } else {
+            // Remove entry from `ongoing_puts`, as we're not part of the IDM group any more
+            let ongoing_puts = mem::replace(&mut self.ongoing_puts, HashMap::new());
+            self.ongoing_puts = ongoing_puts.into_iter()
```
Comment (maqi) [inline src/personas/immutable_data_manager.rs]: will this be more efficient than `self.ongoing_puts.remove(data_name)`  ?

File: src/personas/immutable_data_manager.rs [hunk 2]
```diff
@@ -426,94 +431,174 @@ impl ImmutableDataManager {
...
+
+    fn handle_churn_for_ongoing_gets(&mut self,
+                                     data_name: &XorName,
+                                     close_group: &[XorName])
+                                     -> bool {
+        if let Some(mut metadata) = self.ongoing_gets.get_mut(&data_name) {
+            trace!("Already getting {} - {:?}", data_name, metadata);
+            // Remove any holders which no longer belong in the cache entry
+            metadata.pmid_nodes
+                    .retain(|pmid_node| close_group.contains(pmid_node.name()));
```
Comment (maqi) [inline src/personas/immutable_data_manager.rs]: It was observed in the previous test that there is chance the pmid_node list here to be empty.
As get_mut call will refresh the timestamp of that entry in lru_time_cache, this empty entry will sit in the cache forever.
Though, I am not sure whether this situation will happen with the new code or not?
shall a check shall be considered to be put?

File: src/personas/maid_manager.rs [hunk 3]
```diff
@@ -238,7 +238,9 @@ impl MaidManager {
                              account.put_data(DEFAULT_PAYMENT /* data.payload_size() as u64 */)
                          });
         if let Err(error) = result {
-            trace!("MM responds put_failure of data {}, due to error {:?}", data.name(), error);
+            trace!("MM responds put_failure of data {}, due to error {:?}",
```
Comment (maqi) [inline src/personas/maid_manager.rs]: why clippy complains about this line not the line after?
the next line is even longer.   :(

File: src/personas/immutable_data_manager.rs (hunk 1 above)
Comment (Fraser999) [inline src/personas/immutable_data_manager.rs]: `ongoing_puts` is a `HashMap` indexed by `MessageId`, so we can't use the `data_name` like that.

File: src/personas/maid_manager.rs (hunk 3 above)
Comment (Fraser999) [inline src/personas/maid_manager.rs]: Yeah - I know (although it's rustfmt that does this - not Clippy)  It doesn't seem to handle logging macros at all well.

File: src/personas/immutable_data_manager.rs (hunk 2 above)
Comment (Fraser999) [inline src/personas/immutable_data_manager.rs]: I don't think so - that was a bug previously.  We shouldn't have the situation where `metadata.pmid_nodes` is empty unless all holders churn out before we can replicate.

File: src/personas/immutable_data_manager.rs (hunk 1 above)
Comment (brian-js) [inline src/personas/immutable_data_manager.rs]: What about using the `MessageId` though?

Comment (Fraser999) [inline src/personas/immutable_data_manager.rs]: It's a different ID.  The one passed into this function is derived from the name of the node which caused the churn event, whereas the one in the `ongoing_puts` is the original client-generated one.

**Comment by brian-js** [comment]
Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/367)*

**Comment by brian-js** [comment]
Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/367)*

**Comment by brian-js** [comment]
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/367)*

**Pull request merged**